crates/ruff_python_formatter/src/statement/suite.rs

```diff
@@ -345,6 +327,89 @@ impl FormatRule<Suite, PyFormatContext<'_>> for FormatSuite {
     }
 }
 
+/// Stub files have bespoke rules for empty lines.
+///
+/// These rules are ported from black (preview mode at time of writing) using the stubs test case:
+/// <https://github.com/psf/black/blob/c160e4b7ce30c661ac4f2dfa5038becf1b8c5c33/src/black/lines.py#L576-L744>
+fn stub_file_empty_lines(
+    kind: SuiteKind,
+    preceding: &Stmt,
+    following: &Stmt,
+    preceding_comments: &LeadingDanglingTrailingComments,
+    following_comments: &LeadingDanglingTrailingComments,
+    f: &mut PyFormatter,
+) -> FormatResult<()> {
+    let source = f.context().source();
+    // Preserve the empty line if the definitions are separated by a comment
+    let empty_line_condition = preceding_comments.has_trailing()
+        || following_comments.has_leading()
+        || !stub_suite_can_omit_empty_line(preceding, following, f);
+    match kind {
+        SuiteKind::TopLevel => {
+            if empty_line_condition {
+                empty_line().fmt(f)
+            } else {
+                hard_line_break().fmt(f)
+            }
+        }
+        SuiteKind::Class | SuiteKind::Other | SuiteKind::Function => {
+            if empty_line_condition && lines_after_ignoring_trivia(preceding.end(), source) > 1 {
+                empty_line().fmt(f)
+            } else {
+                hard_line_break().fmt(f)
+            }
+        }
+    }
+}
+
+/// Only a function to compute it lazily
+fn stub_suite_can_omit_empty_line(preceding: &Stmt, following: &Stmt, f: &mut PyFormatter) -> bool {
```

Comment:
Can we make the formatter readonly. The function doesn't seem to call into `fmt`
```suggestion
fn stub_suite_can_omit_empty_line(preceding: &Stmt, following: &Stmt, f: &PyFormatter) -> bool {
```

Reply:
huh thanks good catch, strange that clippy didn't pick that up